### PR TITLE
Enable type checking in CI

### PR DESCRIPTION
**Pull request by pnwamk**
## Summary

Enable Pyright type checking for CompileIQ and tighten several runtime contracts that static analysis exposed.

This adds a `typecheck` dependency group, `pyrightconfig.json`, Pyright CI wiring, and a `make typecheck` target. The implementation keeps the first type-checking gate pragmatic: standard Pyright mode, targeted test relaxations, and a small number of focused suppressions for framework/stub edge cases.

The PR also clarifies the single-objective versus multi-objective API without removing existing behavior. `Search(...)` still works, while new `Search.single_objective(...)` and `Search.multi_objective(...)` constructors make objective intent explicit. `SearchResult.get_best_result()` remains available with preserved behavior, and `pareto_front()` is added as the clearer multi-objective helper.

## Notable Changes

- Add Pyright configuration and CI job.
- Add `make typecheck` and include it in `make validate`.
- Add shared types for objective inputs and objective functions.
- Tighten Pydantic, worker, tracker, validation, and results annotations.
- Preserve `SearchResult.get_best_result()` compatibility while adding `pareto_front()`.
- Add explicit `Search.single_objective(...)` and `Search.multi_objective(...)` constructors.
- Recompute derived pool/cull sizing when an explicit objective constructor changes a `SearchConfiguration` from its default objective count.
- Fix small issues surfaced by typing/tests, including response length checks and multi-config filename generation.

## Validation

- `poetry run pyright`
- `RUFF_CACHE_DIR=/private/tmp/compileiq-ruff-cache poetry run ruff check`
- `make test-unit`
- `make test-integration`

Local results:

- Unit: `191 passed`
- Integration: `66 passed, 13 skipped`

### CODE REVIEW

File: compileiq/ciq.py [hunk 1]
```diff
@@ -145,28 +168,129 @@ class Search(BaseModel):
...
+        config_data["cull_size"] = None
+        derived_config = SearchConfiguration(**config_data)
+        return search_config.pool_size == derived_config.pool_size
+
+    @staticmethod
+    def _uses_derived_cull_size(search_config: SearchConfiguration) -> bool:
+        config_data = search_config.model_dump()
+        config_data["cull_size"] = None
+        derived_config = SearchConfiguration(**config_data)
+        return search_config.cull_size == derived_config.cull_size
```
Comment (vcruzer) [inline compileiq/ciq.py]: I am going to push back on these a little bit, I don't agree with the usage pattern and I am not seeing any benefit from it, it is introducing complexity on something we already did well.

Our usage pattern is based of existing HPO engines on the market. I understand we kept the other way still available, but I don't remember any user issues or requests to have a pattern like this. It feels a bit like over-engineering and introduces a big chunk of code that does not justify it.

One thing I do believe we could do in the future is get rid of `SearchConfiguration` and have it declared inside `Search`. This separation was only done for legacy reasons so `SearchConfiguration` is isolated and contains only the fields required for `core` to start. For this case, I can see us having a similar pattern like this, but it would configure multiple fields for the user at once.

**Review by vcruzer — COMMENTED** [review]
The great majority of these changes are great ! Good work !

File: compileiq/ciq.py (hunk 1 above)
Comment (pnwamk) [reply compileiq/ciq.py]: Thanks, appreciate the feedback.

The explicit `Search.single_objective(...)` / `Search.multi_objective(...)` constructors were meant to make the objective-function contract, especially the return shape, more explicit and less dynamic. After looking at the examples and the surrounding HPO APIs, I agree this is not a big enough win here though. I'm happy to adopt your suggestion and bias toward a smaller API surface that keeps `Search(...)` as the canonical entry point.

If over time the configuration gets complex enough that it becomes difficult to locally reason about what's going on, then maybe we'll want to revisit this part of the API. But that is a concern for future us =)

I'm about to push a commit that removes the added constructors and keeps `Search(...)` as the canonical construction path.

Comment (pnwamk) [reply compileiq/ciq.py]: See [02a9abb](https://github.com/NVIDIA/CompileIQ/pull/10/commits/02a9abbf821d3900addbc776847c7bfb902b3b00)

**Review by vcruzer — APPROVED** [review]
